fahrschein/src/main/java/org/zalando/fahrschein/NakadiClient.java

```diff
@@ -110,17 +110,23 @@ public <T> void publish(String eventName, List<T> events) throws EventPublishing
 
         Response response = null;
         try {
-            eventPublishingHandlers.forEach(requestHandler -> requestHandler.onPublish(eventName, events));
+            eventPublishingHandlers.forEach(handler -> handler.onPublish(eventName, events));
             response = request.execute();
             LOG.debug("Successfully published [{}] events for [{}]", events.size(), eventName);
-            eventPublishingHandlers.forEach(requestHandler -> requestHandler.afterPublish());
         } catch (Throwable t) {
-            eventPublishingHandlers.forEach(requestHandler -> requestHandler.onError(events, t));
+            eventPublishingHandlers.descendingIterator().forEachRemaining(handler -> handler.onError(events, t));
             throw t;
         } finally {
             if(response != null) {
                 response.close();
             }
+            eventPublishingHandlers.descendingIterator().forEachRemaining(handler -> {
+                try {
+                    handler.afterPublish();
```

Comment:
Does it mean that `onError` will get executed only when there are failures during `onPublish`? and it looks like any errors while `afterPublish` will be ignored.

Reply:
that is intentional

Reply:
I mean, the actual reason for `onError` is to get called when errors happen in `request.execute()`

Reply:
One could argue that the eventPublishingHandlers.onPublish() should be guarded by their own try-catch blocks  (we once had an incident in Search caused by ungarded span creation and a bug in Lightstep with baggage items handling)

Reply:
It all comes down to the point where we need to decide how important the handlers are for the publishing process itself. I have the opinion, that it's the responsibility of the handler to manage their error and our try/catch around this would be just the safety net if they don't do this. Because the client is just responsible for executing the handlers and do the publishing.